In [1]:
#import lib
#import thư viện
import pandas as pd
import numpy as np
import matplotlib as plt
import matplotlib.pyplot as plt
import seaborn as sns
print('Completed import lib')

Completed import lib


# Yêu cầu của Robert
- Phải ở khu vực thành thị (urban/metropolitan area).
- Phải ở 1 thành phố mà xếp hạng 75th percentile hoặc cao hơn trong Dữ liệu “metro_startup_ranking” (tức là trong top 25% có rank cao nhất, với rank 1 là cao nhất).
- Phải ở 1 thành phố mà có thấp hơn 50th percentile vể tổng số vụ phạm tội (total number of crime).
- Có chương trình học về lĩnh vực “Computer And Information Sciences And Support Services”.

Ta sử dụng 2 công cụ là Excel và Python để phân tích

# Tiêu chí 1: Thành phố
- Ta thấy dữ liệu trong bảng có tổng cộng 40 rank. Tiêu chí lựa chọn trường của Robert là phải ở 1 thành phố mà xếp hạng 75th percentile hoặc cao hơn trong Dữ liệu “metro_startup_ranking” (tức là trong top 25% có rank cao nhất, với rank 1 là cao nhất). Do Robert có kế hoạch sẽ tìm kiếm 1 công việc thực tập tại 1 startup trong khi theo học.
- Ta sẽ lấy 25% thành phố có rank cao nhất cụ thể là từ rank 1 - 10 trong 40 thành phố trong bảng. (5 Bang: California, Colorado, Florida, New York, Texas)
- Từ các thành phố đạt yêu cầu theo tiêu chí ta tiến hành lọc ra các thành phố trong bảng các vụ phạm tội Crime.

Ta review dữ liệu có trong bảng metro_startup_ranking bằng python

In [2]:
#import dữ liệu 
df_city = pd.read_csv(r'D:\Data Analyst\Learn\MindX\BI12\Final_test\1.Data Wrangling\data\metro_startup_ranking.csv') # / kí tự đặt biệt raw/ rough (để r ở đầu) và file.csv ở cuối

In [3]:
# Hiển thị 5 dòng đầu của bảng metro_startup_ranking 
df_city.head()

Metro Area Code                 Metro Area Name Metro Area Main City  \
0            12060  Atlanta-Sandy Springs-Marietta              Atlanta   
1            12420    Austin-Round Rock-San Marcos               Austin   
2            12580                Baltimore-Towson            Baltimore   
3            14460         Boston-Cambridge-Quincy               Boston   
4            16740    Charlotte-Gastonia-Rock Hill            Charlotte   

               Metro Area States  Startup Rank  
0                        Georgia            13  
1                          Texas             1  
2                       Maryland            17  
3    Massachusetts-New Hampshire            22  
4  North Carolina-South Carolina            25

In [4]:
# Ta sắp xếp Rank các thành phố trong bảng từ cao đến thấp
df_city.sort_values("Startup Rank")

Metro Area Code                           Metro Area Name  \
1             12420              Austin-Round Rock-San Marcos   
18            33100       Miami-Fort Lauderdale-Pompano Beach   
34            41940            San Jose-Sunnyvale-Santa Clara   
17            31100          Los Angeles-Long Beach-Santa Ana   
10            19740                  Denver-Aurora-Broomfield   
33            41860             San Francisco-Oakland-Fremont   
22            35620  New York-Northern New Jersey-Long Island   
12            26420                Houston-Sugar Land-Baytown   
32            41740             San Diego-Carlsbad-San Marcos   
31            41700                 San Antonio-New Braunfels   
16            29820                        Las Vegas-Paradise   
8             18140                                  Columbus   
0             12060            Atlanta-Sandy Springs-Marietta   
25            38060                     Phoenix-Mesa-Glendale   
9             19100               Dallas-Fort Worth-Arlington   
35            42660                   Seattle-Tacoma-Bellevue   
2             12580                          Baltimore-Towson   
29            40140          Riverside-San Bernardino-Ontario   
38            47260       Virginia Beach-Norfolk-Newport News   
37            45300           Tampa-St. Petersburg-Clearwater   
5             16980                 Chicago-Joliet-Naperville   
3             14460                   Boston-Cambridge-Quincy   
30            40900         Sacramento-Arden-Arcade-Roseville   
21            34980  Nashville-Davidson-Murfreesboro-Franklin   
4             16740              Charlotte-Gastonia-Rock Hill   
27            38900              Portland-Vancouver-Hillsboro   
14            27260                              Jacksonville   
13            26900                       Indianapolis-Carmel   
15            28140                               Kansas City   
39            47900           Washington-Arlington-Alexandria   
24            37980            Philadelphia-Camden-Wilmington   
6             17140                     Cincinnati-Middletown   
23            36740                 Orlando-Kissimmee-Sanford   
28            39300         Providence-New Bedford-Fall River   
7             17460                   Cleveland-Elyria-Mentor   
11            19820                    Detroit-Warren-Livonia   
20            33460          Minneapolis-St. Paul-Bloomington   
36            41180                                 St. Louis   
19            33340             Milwaukee-Waukesha-West Allis   
26            38300                                Pittsburgh   

   Metro Area Main City                                  Metro Area States  \
1                Austin                                              Texas   
18                Miami                                            Florida   
34              SanJose                                         California   
17           LosAngeles                                         California   
10               Denver                                           Colorado   
33         SanFrancisco                                         California   
22              NewYork                   New York-New Jersey-Pennsylvania   
12              Houston                                              Texas   
32             SanDiego                                         California   
31           SanAntonio                                              Texas   
16             LasVegas                                             Nevada   
8              Columbus                                               Ohio   
0               Atlanta                                            Georgia   
25              Phoenix                                            Arizona   
9                Dallas                                              Texas   
35              Seattle                                         Washington   
2           

In [5]:
# Lấy ra 25% thành phố có rank cao nhất 
df_city_fil = df_city.loc[df_city['Startup Rank'] < 11 ] 
df_city_fil
# Ta sắp xếp Rank các thành phố trong bảng từ cao đến thấp
df_city_fil.sort_values("Startup Rank")

Metro Area Code                           Metro Area Name  \
1             12420              Austin-Round Rock-San Marcos   
18            33100       Miami-Fort Lauderdale-Pompano Beach   
34            41940            San Jose-Sunnyvale-Santa Clara   
17            31100          Los Angeles-Long Beach-Santa Ana   
10            19740                  Denver-Aurora-Broomfield   
33            41860             San Francisco-Oakland-Fremont   
22            35620  New York-Northern New Jersey-Long Island   
12            26420                Houston-Sugar Land-Baytown   
32            41740             San Diego-Carlsbad-San Marcos   
31            41700                 San Antonio-New Braunfels   

   Metro Area Main City                 Metro Area States  Startup Rank  
1                Austin                             Texas             1  
18                Miami                           Florida             2  
34              SanJose                        California             3  
17           LosAngeles                        California             4  
10               Denver                          Colorado             5  
33         SanFrancisco                        California             6  
22              NewYork  New York-New Jersey-Pennsylvania             7  
12              Houston                             Texas             8  
32             SanDiego                        California             9  
31           SanAntonio                             Texas            10

# Tiêu chí 2: Tình hình tội phạm


In [6]:
#import dữ liệu 
df_crime = pd.read_csv(r'D:\Data Analyst\Learn\MindX\BI12\Final_test\1.Data Wrangling\data\Crime.csv') # / kí tự đặt biệt raw/ rough (để r ở đầu) và file.csv ở cuối

In [7]:
df_crime

MSA  ViolentCrime  Murder  \
0                         Glens Falls, NY M.S.A.4         170.5     0.8   
1                  San German, Puerto Rico M.S.A.          92.1     5.4   
2           Denver-Aurora-Lakewood, CO M.S.A.3, 4         362.3     3.8   
3           Aguadilla-Isabela, Puerto Rico M.S.A.         108.7     4.8   
4    Houston-The Woodlands-Sugar Land, TX M.S.A.3         566.6     6.9   
..                                            ...           ...     ...   
373                        Albuquerque, NM M.S.A.         792.6     6.1   
374                         Alexandria, LA M.S.A.         936.4     4.5   
375                            Hammond, LA M.S.A.         769.4     3.9   
376                             Pueblo, CO M.S.A.         615.7     8.0   
377                            Monroe, LA M.S.A.2        1160.0    15.1   

      Rape  Robbery  AggravatedAssault  PropertyCrime  Burglary   Theft  \
0     67.9      9.5               92.4            NaN     173.7     NaN   
1      4.6     16.1               66.0          420.0     168.9   226.5   
2     63.6     84.8              210.0            NaN       NaN     NaN   
3      3.5     40.5               59.9          710.2     284.5   403.3   
4     37.3    228.9              293.5            NaN     596.8     NaN   
..     ...      ...                ...            ...       ...     ...   
373   63.8    206.7              516.0         4607.8     883.4  3047.6   
374   35.5    120.1              776.3         4565.9    1167.0  3083.7   
375   43.7    141.9              580.0         4770.9    1374.3  3086.2   
376  104.9    114.7              388.2         5190.6    1392.1  3174.8   
377   41.3    122.2              981.3         4701.9    1179.5  3356.0   

     MotorVehicleTheft                                  State         City  
0                 21.3                                     NY  Glens Falls  
1                 24.6         San German, Puerto Rico M.S.A.   San German  
2                388.3                                     CO       Denver  
3                 22.5  Aguadilla-Isabela, Puerto Rico M.S.A.    Aguadilla  
4                367.2                                     TX      Houston  
..                 ...                                    ...          ...  
373              676.9                                     NM  Albuquerque  
374              315.2                                     LA   Alexandria  
375              310.3                                     LA      Hammond  
376              623.7                                     CO       Pueblo  
377              166.3                                     LA       Monroe  

[378 rows x 12 columns]

In [8]:
# thông tin về các trường dữ liệu
df_crime.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 378 entries, 0 to 377
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   MSA                378 non-null    object 
 1   ViolentCrime       377 non-null    float64
 2   Murder             378 non-null    float64
 3   Rape               378 non-null    float64
 4   Robbery            378 non-null    float64
 5   AggravatedAssault  377 non-null    float64
 6   PropertyCrime      372 non-null    float64
 7   Burglary           374 non-null    float64
 8   Theft              375 non-null    float64
 9   MotorVehicleTheft  378 non-null    float64
 10  State              378 non-null    object 
 11  City               373 non-null    object 
dtypes: float64(9), object(3)
memory usage: 35.6+ KB


In [9]:
#thống kê dữ liệu
df_crime.describe()

ViolentCrime      Murder        Rape     Robbery  AggravatedAssault  \
count    377.000000  378.000000  378.000000  378.000000         377.000000   
mean     364.988064    4.574868   41.306614   77.702646         241.319098   
std      181.214645    3.383652   19.506386   53.977181         137.267225   
min       61.600000    0.000000    2.700000    2.300000          25.500000   
25%      236.300000    2.200000   29.000000   36.625000         144.200000   
50%      333.700000    3.800000   37.550000   67.350000         215.700000   
75%      447.700000    6.275000   51.550000  106.050000         304.900000   
max     1160.000000   20.600000  165.600000  351.900000         981.300000   

       PropertyCrime     Burglary        Theft  MotorVehicleTheft  
count     372.000000   374.000000   375.000000         378.000000  
mean     2621.934409   537.622193  1900.392800         191.081746  
std       820.691755   229.176000   576.696896         134.532918  
min       420.000000   159.700000   226.500000          15.700000  
25%      2022.100000   362.725000  1493.900000          97.550000  
50%      2585.250000   503.350000  1854.000000         154.450000  
75%      3207.950000   671.050000  2264.050000         249.650000  
max      5190.600000  1503.700000  3936.400000         718.600000

Trong bảng dữ liệu Crime có tổng cộng 378 bản ghi tương ứng với 378 thành phố. Ta lọc ra 50% số thành phố có tổng số vụ phạm tội thấp nhất để phù hợp với tiêu chí của Robert


In [10]:
# Tính tổng số vụ tội phạm ở các thành phố tại Mỹ
df_crime['TotalofCrime'] = df_crime.sum(axis=1)
# 50% thành phố có có tổng số vụ phạm tội thấp nhất
df_crime.head(189)

MSA  ViolentCrime  Murder  Rape  \
0                         Glens Falls, NY M.S.A.4         170.5     0.8  67.9   
1                  San German, Puerto Rico M.S.A.          92.1     5.4   4.6   
2           Denver-Aurora-Lakewood, CO M.S.A.3, 4         362.3     3.8  63.6   
3           Aguadilla-Isabela, Puerto Rico M.S.A.         108.7     4.8   3.5   
4    Houston-The Woodlands-Sugar Land, TX M.S.A.3         566.6     6.9  37.3   
..                                            ...           ...     ...   ...   
184                     Killeen-Temple, TX M.S.A.         339.5     5.6  67.4   
185                             Dothan, AL M.S.A.         349.6     3.4  35.7   
186                            Midland, TX M.S.A.         322.3     5.4  28.2   
187                          Pocatello, ID M.S.A.         264.3     1.2  31.1   
188                         Morristown, TN M.S.A.         361.8     2.6  31.0   

     Robbery  AggravatedAssault  PropertyCrime  Burglary   Theft  \
0        9.5               92.4            NaN     173.7     NaN   
1       16.1               66.0          420.0     168.9   226.5   
2       84.8              210.0            NaN       NaN     NaN   
3       40.5               59.9          710.2     284.5   403.3   
4      228.9              293.5            NaN     596.8     NaN   
..       ...                ...            ...       ...     ...   
184     63.5              203.0         2559.9     582.0  1855.2   
185     49.9              260.7         2554.0     641.4  1771.8   
186     47.3              241.4         2590.8     524.1  1891.7   
187      8.4              223.6         2665.5     465.2  2052.0   
188     23.3              304.9         2571.9     478.9  1918.1   

     MotorVehicleTheft                                  State         City  \
0                 21.3                                     NY  Glens Falls   
1                 24.6         San German, Puerto Rico M.S.A.   San German   
2                388.3                                     CO       Denver   
3                 22.5  Aguadilla-Isabela, Puerto Rico M.S.A.    Aguadilla   
4                367.2                                     TX      Houston   
..                 ...                                    ...          ...   
184              122.8                                     TX      Killeen   
185              140.8                                     AL       Dothan   
186              174.9                                     TX      Midland   
187              148.3                                     ID    Pocatello   
188              174.8                                     TN   Morristown   

     TotalofCrime  
0           536.1  
1          1024.2  
2          1112.8  
3          1637.9  
4          2097.2  
..            ...  
184        5798.9  
185        5807.3  
186        5826.1  
187        5859.6  
188        5867.3  

[189 rows x 13 columns]

- Từ danh sách ta trên sẽ có được giới hạn cho phép số vụ phạm tội có trong thành phố để phù hợp với tiêu chí chọn trường của Robert là từ 536.1 đến 5867.3 tổng số vụ phạm tội.

In [11]:
# Lọc dữ liệu lấy ra các thành phố phù hơp ở tiêu chí 1
df_crime_fil = df_crime.loc[df_crime['City'].isin(['San Jose','Houston','Denver','San Diego','Austin','Miami','San Antonio','San Francisco','New York','Los Angeles'])]  
df_crime_fil

MSA  ViolentCrime  Murder  \
2                Denver-Aurora-Lakewood, CO M.S.A.3, 4         362.3     3.8   
4         Houston-The Woodlands-Sugar Land, TX M.S.A.3         566.6     6.9   
93                       San Diego-Carlsbad, CA M.S.A.         331.8     2.5   
134          San Jose-Sunnyvale-Santa Clara, CA M.S.A.         251.5     2.3   
155          Los Angeles-Long Beach-Anaheim, CA M.S.A.         431.9     4.9   
175           Los Angeles-Long Beach-Glendale, CA M.D.         495.7     5.8   
192                      Austin-Round Rock, TX M.S.A.2         287.7     2.2   
302   Miami-Fort Lauderdale-West Palm Beach, FL M.S.A.         510.1     6.6   
333           San Francisco-Oakland-Hayward, CA M.S.A.         484.9     5.3   
350                 Miami-Miami Beach-Kendall, FL M.D.         610.3     8.0   
353               San Antonio-New Braunfels, TX M.S.A.         436.9     5.4   
363  San Francisco-Redwood City-South San Francisco...         520.7     4.1   

     Rape  Robbery  AggravatedAssault  PropertyCrime  Burglary   Theft  \
2    63.6     84.8              210.0            NaN       NaN     NaN   
4    37.3    228.9              293.5            NaN     596.8     NaN   
93   33.3     84.9              211.1         1922.1     325.4  1296.6   
134  29.6     80.1              139.6         2316.7     438.9  1406.3   
155  35.1    149.1              242.8         2308.1     418.5  1496.1   
175  38.4    173.7              277.8         2357.7     445.9  1480.8   
192  44.9     61.2              179.4         2672.8     415.7  2096.8   
302  30.9    156.8              315.8         3285.8     536.1  2470.0   
333  30.9    235.2              213.5         3600.5     517.6  2451.3   
350  27.8    197.5              377.0         3670.5     534.9  2811.8   
353  59.0     94.2              278.3         3952.6     657.4  2982.9   
363  34.7    256.4              225.5         4259.6     524.4  3190.4   

     MotorVehicleTheft State           City  TotalofCrime  
2                388.3    CO         Denver        1112.8  
4                367.2    TX        Houston        2097.2  
93               300.0    CA      San Diego        4507.7  
134              471.4    CA       San Jose        5136.4  
155              393.5    CA    Los Angeles        5480.0  
175              431.1    CA    Los Angeles        5706.9  
192              160.2    TX         Austin        5920.9  
302              279.7    FL          Miami        7591.8  
333              631.7    CA  San Francisco        8170.9  
350              323.9    FL          Miami        8561.7  
353              312.4    TX    San Antonio        8779.1  
363              544.9    CA  San Francisco        9560.7

In [12]:
# Lọc ra giới hạn cho phép số vụ phạm tội có trong thành phố để phù hợp với tiêu chí chọn trường của Robert là dưới 5867.3 tổng số vụ phạm tội
df_crime_fil2 = df_crime_fil.loc[df_crime_fil['TotalofCrime'] < 5867.3 ]  
df_crime_fil2

MSA  ViolentCrime  Murder  Rape  \
2           Denver-Aurora-Lakewood, CO M.S.A.3, 4         362.3     3.8  63.6   
4    Houston-The Woodlands-Sugar Land, TX M.S.A.3         566.6     6.9  37.3   
93                  San Diego-Carlsbad, CA M.S.A.         331.8     2.5  33.3   
134     San Jose-Sunnyvale-Santa Clara, CA M.S.A.         251.5     2.3  29.6   
155     Los Angeles-Long Beach-Anaheim, CA M.S.A.         431.9     4.9  35.1   
175      Los Angeles-Long Beach-Glendale, CA M.D.         495.7     5.8  38.4   

     Robbery  AggravatedAssault  PropertyCrime  Burglary   Theft  \
2       84.8              210.0            NaN       NaN     NaN   
4      228.9              293.5            NaN     596.8     NaN   
93      84.9              211.1         1922.1     325.4  1296.6   
134     80.1              139.6         2316.7     438.9  1406.3   
155    149.1              242.8         2308.1     418.5  1496.1   
175    173.7              277.8         2357.7     445.9  1480.8   

     MotorVehicleTheft State         City  TotalofCrime  
2                388.3    CO       Denver        1112.8  
4                367.2    TX      Houston        2097.2  
93               300.0    CA    San Diego        4507.7  
134              471.4    CA     San Jose        5136.4  
155              393.5    CA  Los Angeles        5480.0  
175              431.1    CA  Los Angeles        5706.9

In [13]:
#kiểm tra dữ liệu có null ko, mất k - xử lí dữ liệu
df_crime_fil2.isna().sum()

MSA                  0
ViolentCrime         0
Murder               0
Rape                 0
Robbery              0
AggravatedAssault    0
PropertyCrime        2
Burglary             1
Theft                2
MotorVehicleTheft    0
State                0
City                 0
TotalofCrime         0
dtype: int64

In [14]:
#Làm sạch dữ liệu - bỏ các thành phố ko có dữ liệu
df_crime_fil3 = df_crime_fil2.dropna(subset = ['PropertyCrime','PropertyCrime','Theft'])
df_crime_fil3

MSA  ViolentCrime  Murder  Rape  \
93               San Diego-Carlsbad, CA M.S.A.         331.8     2.5  33.3   
134  San Jose-Sunnyvale-Santa Clara, CA M.S.A.         251.5     2.3  29.6   
155  Los Angeles-Long Beach-Anaheim, CA M.S.A.         431.9     4.9  35.1   
175   Los Angeles-Long Beach-Glendale, CA M.D.         495.7     5.8  38.4   

     Robbery  AggravatedAssault  PropertyCrime  Burglary   Theft  \
93      84.9              211.1         1922.1     325.4  1296.6   
134     80.1              139.6         2316.7     438.9  1406.3   
155    149.1              242.8         2308.1     418.5  1496.1   
175    173.7              277.8         2357.7     445.9  1480.8   

     MotorVehicleTheft State         City  TotalofCrime  
93               300.0    CA    San Diego        4507.7  
134              471.4    CA     San Jose        5136.4  
155              393.5    CA  Los Angeles        5480.0  
175              431.1    CA  Los Angeles        5706.9

- Ở bảng trên ta có thể thấy Denver và Houston bị mất dữ liệu ở 3 cột PropertyCrime, Burglary, Theft.
- New York không có bản ghi trong bảng dữ liệu Crime 

Ta sẽ tạm loại bớt 3/6 thành phố trên và chỉ còn San Jose, Los Angeles và San Diego	

# Tiêu chí 3: Trường đại học

Trước hết ta đọc Dictionary for University_info để xác định các trường có chương trình học "Computer And Information Sciences And Support Services" và ở khu vực thành thị. Ta giữ lại các cột liên qua và bỏ đi các cột thừa
## 1. Các trường ở thành thị
- Đầu tiên ta lọc ra các trường ở 3 khu vực là **San Jose, Los Angeles và San Diego**
- Trong bảng **University_info** ta có cột LOCALE. Trong cột này ta giữ lại các trường đại học có giá trị là **11** (City: Large - population of 250,000 or more) đây là các trường đại học nằm ở các thành phố lớn.

## 2. Chương trình học
- Ta đã xác định được các trường có chương trình học **"Computer And Information Sciences And Support Services"** thường sẽ có dữ liệu trong cột **PCIP11** (Percentage of degrees awarded in Computer And Information Sciences And Support Services) của bảng dữ liệu University_info. Đây là các trường có đào tạo và phát bằng cấp cho Ngành học - Computer And Information Sciences And Support Services mà Robert muốn theo học.
- Tuy nhiên có một số trường đại học cột **PCIP11 giá trị bằng 0** do đó ta sẽ dựa vào **5 cột (CIP11CERT1, CIP11CERT2, CIP11ASSOC, CIP11CERT4, CIP11BACHL) để chọn thêm trường cho Robert**. Đây là các cột thể hiện các bằng cấp chứng chỉ mà trường đó đào tạo với các giá trị 0, 1 ,2 (0: Ko có chương trình, 1: Có chương trình học, 2: Đào tạo từ xa).
- Ta giữ các cột có liên quan đến tiêu chí chọn trường của Robert và loại bỏ bớt các cột không cần thiết 
- Ta xóa các cột không cần thiết trong bảng dữ liệu và chỉ để lại các dữ liệu có liên quan đến ngành học để import vào Python

In [15]:
#import dữ liệu các trường đại học
df_university = pd.read_csv(r'D:\Data Analyst\Learn\MindX\BI12\Final_test\1.Data Wrangling\data\University_info.csv') # / kí tự đặt biệt raw/ rough (để r ở đầu) và file.csv ở cuối

C:\Users\ADMIN\AppData\Roaming\Python\Python38\site-packages\IPython\core\interactiveshell.py:3012: DtypeWarning: Columns (7,9,1427,1542,1561,1575,1725,1726,1727,1728) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


In [16]:
# 5 dòng đầu của dữ liệu
df_university.head()

UNITID    OPEID  opeid6                               INSTNM        CITY  \
0  100654   100200    1002             Alabama A & M University      Normal   
1  100663   105200    1052  University of Alabama at Birmingham  Birmingham   
2  100690  2503400   25034                   Amridge University  Montgomery   
3  100706   105500    1055  University of Alabama in Huntsville  Huntsville   
4  100724   100500    1005             Alabama State University  Montgomery   

  STABBR         ZIP                                       AccredAgency  \
0     AL       35762  Southern Association of Colleges and Schools C...   
1     AL  35294-0110  Southern Association of Colleges and Schools C...   
2     AL  36117-3553  Southern Association of Colleges and Schools C...   
3     AL       35899  Southern Association of Colleges and Schools C...   
4     AL  36104-0271  Southern Association of Colleges and Schools C...   

                          INSTURL  \
0                   www.aamu.edu/   
1                     www.uab.edu   
2       www.amridgeuniversity.edu   
3                     www.uah.edu   
4  www.alasu.edu/email/index.aspx   

                                              NPCURL  ...  \
0     galileo.aamu.edu/netpricecalculator/npcalc.htm  ...   
1    www.collegeportraits.org/AL/UAB/estimator/agree  ...   
2  tcc.noellevitz.com/(S(miwoihs5stz5cpyifh4nczu0...  ...   
3                                    finaid.uah.edu/  ...   
4  www.alasu.edu/cost-aid/forms/calculator/index....  ...   

   PELL_RPY_3YR_RT_SUPP  NOPELL_RPY_3YR_RT_SUPP  FEMALE_RPY_3YR_RT_SUPP  \
0        0.443586698337           0.45251396648          0.484818805093   
1        0.732297852901          0.795454545455          0.760521885522   
2        0.600877192982                   0.775          0.610465116279   
3        0.738341968912           0.85140562249          0.768378650554   
4        0.320016820858          0.380829015544          0.364423717522   

   MALE_RPY_3YR_RT_SUPP  FIRSTGEN_RPY_3YR_RT_SUPP  \
0                 0.406            0.423580786026   
1        0.755344418052            0.751552795031   
2        0.686567164179            0.632653061224   
3        0.798440979955            0.778156996587   
4        0.286530612245            0.315737051793   

   NOTFIRSTGEN_RPY_3YR_RT_SUPP  C150_L4_POOLED_SUPP  C150_4_POOLED_SUPP  \
0               0.442452830189                  NaN    0.30871831265508   
1               0.765074770863                  NaN    0.50854980789381   
2                         0.65                  NaN   PrivacySuppressed   
3               0.791666666667                  NaN    0.47821134615384   
4               0.334633385335                  NaN    0.25748195545488   

   C200_L4_POOLED_SUPP  C200_4_POOLED_SUPP  
0                  NaN    0.35168556430446  
1                  NaN     0.4730923834537  
2                  NaN   PrivacySuppressed  
3                  NaN    0.49412939796716  
4                  NaN    0.29365162671232  

[5 rows x 1729 columns]

In [17]:
# Lấy ra những cột cần thiết cho phân tích
df_university_filter = df_university.filter(["UNITID", "INSTNM", "CITY",'st_fips','region','LOCALE','PCIP11','CIP11CERT1','CIP11CERT2','CIP11ASSOC','CIP11CERT4','CIP11BACHL'])
df_university_filter.head()

UNITID                               INSTNM        CITY  st_fips  region  \
0  100654             Alabama A & M University      Normal        1       5   
1  100663  University of Alabama at Birmingham  Birmingham        1       5   
2  100690                   Amridge University  Montgomery        1       5   
3  100706  University of Alabama in Huntsville  Huntsville        1       5   
4  100724             Alabama State University  Montgomery        1       5   

   LOCALE  PCIP11  CIP11CERT1  CIP11CERT2  CIP11ASSOC  CIP11CERT4  CIP11BACHL  
0    12.0  0.0348         0.0         0.0         0.0         0.0         2.0  
1    12.0  0.0099         0.0         1.0         0.0         1.0         2.0  
2    12.0  0.0411         0.0         0.0         2.0         0.0         2.0  
3    12.0  0.0273         0.0         0.0         0.0         0.0         2.0  
4    12.0  0.0450         0.0         0.0         0.0         0.0         2.0

In [18]:
# lọc ra các trường đại học tương ứng với 3 thành phố San Jose, Los Angeles và San Diego trong tiêu chí
df_uni = df_university_filter.loc[df_university_filter['CITY'].isin(['San Jose','San Diego','Los Angeles'])]  
# lọc ra các trường đại học ở thành thị với giá trị LOCALE = 11
df_uni2 = df_uni.loc[df_uni['LOCALE'] == 11 ] 
# Lọc ra các trường có đào tạo ngành học của Robert (các cột có giá trị > 0: có chương trình thuộc ngành Computer And Information Sciences And Support Services ) 
df_uni3 = df_uni2.loc[(df_uni2['CIP11CERT1'] > 0) | (df_uni2['CIP11CERT2'] > 0) |(df_uni2['CIP11ASSOC'] > 0) |(df_uni2['CIP11CERT4'] > 0) |(df_uni2['CIP11BACHL'] > 0)]
df_uni3

UNITID                                             INSTNM         CITY  \
216   109730             Associated Technical College-San Diego    San Diego   
247   110592            California State University-Los Angeles  Los Angeles   
253   110662               University of California-Los Angeles  Los Angeles   
262   110945                       California College San Diego    San Diego   
297   112446                                 Coleman University    San Diego   
326   114266                           Evergreen Valley College     San Jose   
376   117113  The Art Institute of California-Argosy Univers...    San Diego   
393   117724                Los Angeles Trade Technical College  Los Angeles   
396   117788                           Los Angeles City College  Los Angeles   
401   117946                        Loyola Marymount University  Los Angeles   
453   120254                                 Occidental College  Los Angeles   
465   120838                          Pacific States University  Los Angeles   
474   121275                            Platt College-San Diego    San Diego   
475   121309                     Point Loma Nazarene University    San Diego   
495   122339                             San Diego City College    San Diego   
497   122375                             San Diego Mesa College    San Diego   
498   122384                          San Diego Miramar College    San Diego   
499   122409                         San Diego State University    San Diego   
500   122436                            University of San Diego    San Diego   
511   122746                              San Jose City College     San Jose   
512   122755                          San Jose State University     San Jose   
515   122843                       Westwood College-Los Angeles  Los Angeles   
548   123961                  University of Southern California  Los Angeles   
4573  368780         Los Angeles ORT College-Los Angeles Campus  Los Angeles   
4612  372222             University of Phoenix-San Diego Campus    San Diego   
4874  406006              University of Phoenix-Bay Area Campus     San Jose   
5404  437936             Carrington College California-San Jose     San Jose   
7259  481234                       Advanced Computing Institute  Los Angeles   

      st_fips  region  LOCALE  PCIP11  CIP11CERT1  CIP11CERT2  CIP11ASSOC  \
216         6       8    11.0  0.1809         1.0         0.0         0.0   
247         6       8    11.0  0.0297         0.0         0.0         0.0   
253         6       8    11.0  0.0156         0.0         0.0         0.0   
262         6       8    11.0  0.0907         0.0         0.0         1.0   
297         6       8    11.0  1.0000         0.0         1.0         1.0   
326         6       8    11.0  0.0000         1.0         0.0         2.0   
376         6       8    11.0  0.1202         0.0         1.0         1.0   
393         6       8    11.0  0.0082         1.0         1.0         1.0   
396         6       8    11.0  0.0062         1.0         1.0         1.0   
401         6       8    11.0  0.0039         0.0         0.0         0.0   
453         6       8    11.0  0.0000         0.0         0.0         0.0   
465         6       8    11.0  0.0000         0.0         0.0         0.0   
474         6       8    11.0  0.0359         1.0         1.0         1.0   
475         6       8    11.0  0.0071         0.0         0.0         0.0   
495         6       8    11.0  0.0062         1.0         1.0         1.0   
497         6       8    11.0  0.0295         1.0         1.0         1.0   
498         6       8    11.0  0.0103         1.0         1.0         1.0   
499         6       8    11.0  0.0109         0.0         0.0         0.0   
500         6       8    11.0  0.0045         0.0         0.0         0.0   
511         6       8    11.0  0.0159         1.0         1.0         2.0   
512         6       8    11.0  0.0140         0.0         0.0       

In [19]:
#Thông tin dữ liệu
df_uni3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 28 entries, 216 to 7259
Data columns (total 12 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   UNITID      28 non-null     int64  
 1   INSTNM      28 non-null     object 
 2   CITY        28 non-null     object 
 3   st_fips     28 non-null     int64  
 4   region      28 non-null     int64  
 5   LOCALE      28 non-null     float64
 6   PCIP11      28 non-null     float64
 7   CIP11CERT1  28 non-null     float64
 8   CIP11CERT2  28 non-null     float64
 9   CIP11ASSOC  28 non-null     float64
 10  CIP11CERT4  28 non-null     float64
 11  CIP11BACHL  28 non-null     float64
dtypes: float64(7), int64(3), object(2)
memory usage: 2.8+ KB


In [20]:
#Xem dữ liệu có bị null không
df_uni3.isnull().sum()

UNITID        0
INSTNM        0
CITY          0
st_fips       0
region        0
LOCALE        0
PCIP11        0
CIP11CERT1    0
CIP11CERT2    0
CIP11ASSOC    0
CIP11CERT4    0
CIP11BACHL    0
dtype: int64

In [21]:
# Có bảo nhiêu trường đại học
df_uni3.shape

(28, 12)

Như vậy có tổng cộng **28 trường đại học trong 3 thành phố là San Jose, Los Angeles và San Diego (Tất cả đều ở bang California)** đạt tiêu chí của Robert

## Tạo bảng danh sách các trường đại học phù hợp với tiêu chí của Robert

In [22]:
# Danh sách các trường đại học phù hợp với tiêu chí của Robert
df = df_uni3.filter(["CITY", "INSTNM", "UNITID"])
df

CITY                                             INSTNM  UNITID
216     San Diego             Associated Technical College-San Diego  109730
247   Los Angeles            California State University-Los Angeles  110592
253   Los Angeles               University of California-Los Angeles  110662
262     San Diego                       California College San Diego  110945
297     San Diego                                 Coleman University  112446
326      San Jose                           Evergreen Valley College  114266
376     San Diego  The Art Institute of California-Argosy Univers...  117113
393   Los Angeles                Los Angeles Trade Technical College  117724
396   Los Angeles                           Los Angeles City College  117788
401   Los Angeles                        Loyola Marymount University  117946
453   Los Angeles                                 Occidental College  120254
465   Los Angeles                          Pacific States University  120838
474     San Diego                            Platt College-San Diego  121275
475     San Diego                     Point Loma Nazarene University  121309
495     San Diego                             San Diego City College  122339
497     San Diego                             San Diego Mesa College  122375
498     San Diego                          San Diego Miramar College  122384
499     San Diego                         San Diego State University  122409
500     San Diego                            University of San Diego  122436
511      San Jose                              San Jose City College  122746
512      San Jose                          San Jose State University  122755
515   Los Angeles                       Westwood College-Los Angeles  122843
548   Los Angeles                  University of Southern California  123961
4573  Los Angeles         Los Angeles ORT College-Los Angeles Campus  368780
4612    San Diego             University of Phoenix-San Diego Campus  372222
4874     San Jose              University of Phoenix-Bay Area Campus  406006
5404     San Jose             Carrington College California-San Jose  437936
7259  Los Angeles                       Advanced Computing Institute  481234

In [23]:
# Đổi tên các cột trong các bảng để join dữ liệu
df_crime_rename = df_crime_fil3.rename(columns ={'City':'CITY'})
# Lấy ra những cột cần thiết
crimee = df_crime_rename.filter(["CITY", "TotalofCrime"])
crime = crimee.loc[(crimee['TotalofCrime'] < 5706.9)]
crime #Tổng số vụ tội phạm trong các thành phố mà Robert sẽ theo học

CITY  TotalofCrime
93     San Diego        4507.7
134     San Jose        5136.4
155  Los Angeles        5480.0

In [24]:
# Đổi tên cột trong bảng xếp hạng Startup Rank của các thành phố
df_city_fil2 = df_city_fil.rename(columns ={'Metro Area Main City':'CITY'})

# lọc ra thành phố phù hợp với tiêu chí
city_filter = df_city_fil2.filter(["CITY", "Metro Area States",'Startup Rank'])
city = city_filter.loc[(city_filter['CITY'] == 'LosAngeles') | (city_filter['CITY'] == 'SanDiego') | (city_filter['CITY'] == 'SanJose')]                      

# Đổi lại định dạng cho thành phố
city["CITY"].replace({"LosAngeles": "Los Angeles", "SanDiego": "San Diego","SanJose": "San Jose" }, inplace=True)
city

C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\series.py:4509: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().replace(


CITY Metro Area States  Startup Rank
17  Los Angeles        California             4
32    San Diego        California             9
34     San Jose        California             3

In [25]:
# combine data
robert = pd.merge(df,crime,on='CITY', validate='many_to_many')
robert_university = pd.merge(robert,city,on='CITY', validate='many_to_many')
# Bảng tổng hợp các trường đại học theo tiêu chí của Robert
robert_university

CITY                                             INSTNM  UNITID  \
0     San Diego             Associated Technical College-San Diego  109730   
1     San Diego                       California College San Diego  110945   
2     San Diego                                 Coleman University  112446   
3     San Diego  The Art Institute of California-Argosy Univers...  117113   
4     San Diego                            Platt College-San Diego  121275   
5     San Diego                     Point Loma Nazarene University  121309   
6     San Diego                             San Diego City College  122339   
7     San Diego                             San Diego Mesa College  122375   
8     San Diego                          San Diego Miramar College  122384   
9     San Diego                         San Diego State University  122409   
10    San Diego                            University of San Diego  122436   
11    San Diego             University of Phoenix-San Diego Campus  372222   
12  Los Angeles            California State University-Los Angeles  110592   
13  Los Angeles               University of California-Los Angeles  110662   
14  Los Angeles                Los Angeles Trade Technical College  117724   
15  Los Angeles                           Los Angeles City College  117788   
16  Los Angeles                        Loyola Marymount University  117946   
17  Los Angeles                                 Occidental College  120254   
18  Los Angeles                          Pacific States University  120838   
19  Los Angeles                       Westwood College-Los Angeles  122843   
20  Los Angeles                  University of Southern California  123961   
21  Los Angeles         Los Angeles ORT College-Los Angeles Campus  368780   
22  Los Angeles                       Advanced Computing Institute  481234   
23     San Jose                           Evergreen Valley College  114266   
24     San Jose                              San Jose City College  122746   
25     San Jose                          San Jose State University  122755   
26     San Jose              University of Phoenix-Bay Area Campus  406006   
27     San Jose             Carrington College California-San Jose  437936   

    TotalofCrime Metro Area States  Startup Rank  
0         4507.7        California             9  
1         4507.7        California             9  
2         4507.7        California             9  
3         4507.7        California             9  
4         4507.7        California             9  
5         4507.7        California             9  
6         4507.7        California             9  
7         4507.7        California             9  
8         4507.7        California             9  
9         4507.7        California             9  
10        4507.7        California             9  
11        4507.7        California             9  
12        5480.0        California             4  
13        5480.0        California             4  
14        5480.0        California             4  
15        5480.0        California             4  
16        5480.0        California             4  
17        5480.0        California             4  
18        5480.0        California             4  
19        5480.0        California             4  
20        5480.0        California             4  
21        5480.0        California             4  
22        5480.0        California             4  
23        5136.4        California             3  
24        5136.4        California             3  
25        5136.4        California             3  
26        5136.4        California             3  
27        5136.4        California             3

**--> Như vậy có tổng cộng 28 trường đại học ở 3 thành phố San Diego, Los Angeles, San Jose	ở bang California phù hợp với tiêu chí của Robert**